### ISSUE TITLE
Error installing rclone from GitHub

### ISSUE THREAD

**Issue by dbareiro** [ncw/rclone#67]
Hi, Nick.

Maybe you can help me with this error. I do not think it's a problem of ```rclone``` itself, but some dependency that is not installed. I'm using Debian GNU/Linux Jessie.

```
# aptitude install golang git gcc
```

Then:

```
# go install github.com/ncw/rclone
# github.com/ncw/rclone/googlecloudstorage
go/src/github.com/ncw/rclone/googlecloudstorage/googlecloudstorage.go:45: undefined: storage.DevstorageFull_controlScope
```

Maybe I'm missing to install some Go library?

Thanks in advance.

Best regards,
Daniel

**Comment by ncw** [comment]
You probably have an old version of the `github.com/ncw/rclone/googlecloudstorage` library - this changed recently.

Try 

```
cd $GOPATH/src/github.com/ncw/rclone
go get -u -f -v ./...
```

and that should update all the dependencies.

**Comment by dbareiro** [comment]
Hi, Nick. Thanks for your reply.

I tried what you suggested me, but I got the following error:

```
root@icecast:~# cd $GOPATH/src/github.com/ncw/rclone
root@icecast:~/go/src/github.com/ncw/rclone# go get -u -f -v ./...
flag provided but not defined: -f
usage: get [-d] [-fix] [-t] [-u] [build flags] [packages]

Get downloads and installs the packages named by the import paths,
along with their dependencies.

The -d flag instructs get to stop after downloading the packages; that is,
it instructs get not to install the packages.

The -fix flag instructs get to run the fix tool on the downloaded packages
before resolving dependencies or building the code.

The -t flag instructs get to also download the packages required to build
the tests for the specified packages.

The -u flag instructs get to use the network to update the named packages
and their dependencies.  By default, get uses the network to check out
missing packages but does not use it to look for updates to existing packages.

Get also accepts build flags to control the installation. See 'go help build'.

When checking out or updating a package, get looks for a branch or tag
that matches the locally installed version of Go. The most important
rule is that if the local installation is running version "go1", get
searches for a branch or tag named "go1". If no such version exists it
retrieves the most recent version of the package.

For more about specifying packages, see 'go help packages'.

For more about how 'go get' finds source code to
download, see 'go help importpath'.

See also: go build, go install, go clean.
```
Perhaps the parameter should be ```fix```?

Although perhaps the name of the parameter depends on the installed version:

```
root@icecast:~/go/src/github.com/ncw/rclone# aptitude show golang
Paquete: golang                                  
Nuevo: sí
Estado: instalado
Instalado automáticamente: no
Versión: 2:1.3.3-1
Prioridad: opcional
Sección: devel
Desarrollador: Go Compiler Team <pkg-golang-devel@lists.alioth.debian.org>
Arquitectura: all
Tamaño sin comprimir: 91,1 k
Depende de: golang-go (>= 2:1.3.3-1), golang-src (>= 2:1.3.3-1), golang-doc (>= 2:1.3.3-1)
Tiene conflictos con: golang-weekly
Reemplaza: golang-weekly
Proporciona: golang-weekly
Descripción: Go programming language compiler - metapackage
 The Go programming language is an open source project to make programmers more productive. Go is expressive, concise, clean, and efficient. Its
 concurrency mechanisms make it easy to write programs that get the most out of multicore and networked machines, while its novel type system enables
 flexible and modular program construction. Go compiles quickly to machine code yet has the convenience of garbage collection and the power of run-time
 reflection. It's a fast, statically typed, compiled language that feels like a dynamically typed, interpreted language. 
 
 This package is a metapackage that, when installed, guarantees that (most of) a full Go development environment is installed.
Página principal: http://golang.org/
```

Best regards,
Daniel

**Comment by ncw** [comment]
Ah, you are using go 1.3 - in that case leave the `-f` flag off `go get -u -v ./...` should do it.

Alternatively install the latest go which is really easy using [godeb](https://github.com/niemeyer/godeb) which will download the latest source release for you and build it into a `.deb`.

**Comment by dbareiro** [comment]
Hi, Nick.

Thank you very much. That worked:

```
root@icecast:~/go/src/github.com/ncw/rclone# go get -u -v ./...
github.com/ncw/rclone (download)
code.google.com/p/goauth2 (download)
github.com/Unknwon/goconfig (download)
github.com/mreiferson/go-httpclient (download)
github.com/ncw/goamz (download)
github.com/ogier/pflag (download)
github.com/tsenart/tb (download)
golang.org/x/net (download)
google.golang.org/api (download)
github.com/stacktic/dropbox (download)
golang.org/x/oauth2 (download)
github.com/ncw/swift (download)
github.com/ncw/rclone/fs
github.com/ncw/rclone/googleauth
github.com/ncw/rclone/drive
golang.org/x/oauth2
github.com/stacktic/dropbox
github.com/ncw/rclone/dropbox
google.golang.org/api/storage/v1
github.com/ncw/rclone/googlecloudstorage
github.com/ncw/rclone/local
github.com/ncw/swift
github.com/ncw/rclone/s3
github.com/ncw/rclone/swift
github.com/ncw/rclone
github.com/ncw/rclone/fstest
github.com/ncw/rclone/fstest/fstests
```
Thank you also for your suggestion about ```godeb```. I will have that in mind.

Some questions:

* Is there any way to make the binary during compilation is placed in a standard path of the system (for example, ```/usr/local/bin/)```?
* What command should be used to generate the documentation? This would also be ideal to be in a standard system path for man pages.

Thanks again.

Best regards,
Daniel

**Comment by ncw** [comment]
What I do is make sure that $GOPATH/bin is on the path, eg

```
$ echo $GOPATH
/home/ncw/Code/Go

$ echo $PATH
/home/ncw/Code/Go/bin/:/home/ncw/bin:etc...
```

I don't think there is a way of placing the binary directly in `/usr/local/bin` though.

You can make the man page with `make rclone.1` which uses `pandoc`. You'll have to copy it to a standard place.

It would be quite easy to package up rclone into a .deb - I've never quite got round to it though!

**Comment by dbareiro** [comment]
Hi, Nick.

Thanks for your reply.

> What I do is make sure that $GOPATH/bin is on the path, eg
>
> ```
> $ echo $GOPATH
> /home/ncw/Code/Go
> 
> $ echo $PATH
> /home/ncw/Code/Go/bin/:/home/ncw/bin:etc...
> ```
> I don't think there is a way of placing the binary directly in /usr/local/bin though.

Yes, now I have this configuration in my ```.profile```:

```
if [ -d "$HOME/go" ] ; then
    export GOPATH="$HOME/go"
fi

if [ -d "$HOME/go/bin" ] ; then
    PATH="$HOME/go/bin:$PATH"
fi
```

Although my idea was to have the executable on a general system path. I guess if there is no other way, I'll have to manually copy it to ```/usr/local/bin```. No problem. Just I was curious to know if there was any way to get the binary in a specific path :-)

Update: I think we can do so by modifying the ```Makefile```:

```
install: rclone
        #install -d ${DESTDIR}/usr/bin
        #install -t ${DESTDIR}/usr/bin ${GOPATH}/bin/rclone
        install -d ${DESTDIR}
        install -t ${DESTDIR} ${GOPATH}/bin/rclone
```

Then:

```
# make install DESTDIR=/usr/local/bin
go version go1.3.3 linux/386
go install -v ./...
github.com/ncw/rclone
#install -d /usr/local/bin/usr/bin
#install -t /usr/local/bin/usr/bin /root/go/bin/rclone
install -d /usr/local/bin
install -t /usr/local/bin /root/go/bin/rclone
```

```
root@icecast:~/go/src/github.com/ncw/rclone# ll ~/go/bin/ /usr/local/bin/
/root/go/bin/:
total 8316
-rwxr-xr-x 1 root root 8503704 jun  3 09:58 rclone

/usr/local/bin/:
total 8316
-rwxr-xr-x 1 root root 8503704 jun  3 09:58 rclone
```

And if you want to install only on the ```$GOPATH``` directory simply use:

```
# make rclone
```

I'm not sure if there will be a more uniform do so using ```make install``` with a default value given that if ```DESTDIR``` is not specified, an error is returned.

> You can make the man page with ```make rclone.1``` which uses ```pandoc```. You'll have to copy it to a standard place.

Yes. Yesterday I was just looking at this line on the ```Makefile``` and I managed to generate ```rclone.1``` writing that line in particular (previously installing the ```pandoc``` package):

```
rclone.1:       README.md
        pandoc -s --from markdown --to man README.md -o rclone.1
```

A detail to mention: when generate the ```rclone.1``` file and try to ```man ./rclone.1```, it looks something like this:

```
Rclone
       [IMAGE: Logo (http://rclone.org/img/rclone-120x120.png)] (http://rclone.org/)
```
It is a minor detail, but maybe there is some way to remove the reference image in the manual page.

I think it could be interesting and beneficial to many to include in the installation documentation some things we mentioned in this thread.

> It would be quite easy to package up rclone into a .deb - I've never quite got round to it though!

Perhaps by talking to some Debian developer, they are interested in including your software in the distribution. This way more people could know it, use and contribute to the development or reporting bugs, and everyone would benefit.

Moreover, in issue #66 I had answered some of your questions and also I showed you and consulted some other things. Not sure if you've received it, since you already had closed the issue; but I would appreciate if you could see it.

Best regards,
Daniel

**Comment by ncw** [comment]
In 5835f15f21178a2c8f9a6f8e7e4aaff372764410 some linux install instructions were added.

I've also improved the man page somewhat, but that IMAGE thing is still there!

I would like to package this up as a deb some time, but I don't have time at the moment.

I'm going to close this ticket as it has now gone off topic.

If you'd like to open specific focused tickets about changes you'd like I'd be grateful, or even better send some pull requests.

Thanks for your contributions

Nick
